#This script is pulled together from various sources, including Dr. Dworkin's in-class tutorial, Garrett's differential expression analysis script, and the DESeq2 vignette/manual/other resources I found online. 

#Load in the packages required for the analysis

library(DESeq2)
library (readr)
library (gplots)
library("RColorBrewer")

#Set the working directory to the folder where the counts are stored

setwd("C:/Users/rowbr/OneDrive/Documents/Counts")

#Check that you're actually in the right directory

getwd()

#Check that all the files we expect to be in the working directory are actually there

dir()

#Naming all of our samples manually
#Since we have samples from 2 data sets that are named differently, I have to rename them in a consistent format.
#Just a note for clarity: the count files with treatment label EV will be renamed with DC3000 as the treatment. This treatment was inoculation with P. syringae pv. tomato DC3000 carrying an empty vector (EV) rather than a plasmid with an effector protein like in the other treatments.
#Names follow the format: treatment_timepoint_replicate_dataset.

samples <- c("mock_12h_01_eti",
             "mock_12h_02_eti",
             "mock_12h_03_eti",
             "mock_24h_01_eti",
             "mock_24h_02_eti",
             "mock_24h_03_eti",
             "mock_6h_01_eti",
             "mock_6h_02_eti",
             "mock_6h_03_eti",
             "DC3000_12h_01_eti",
             "DC3000_12h_02_eti",
             "DC3000_12h_03_eti",
             "DC3000_24h_01_eti",
             "DC3000_24h_02_eti",
             "DC3000_24h_03_eti",
             "DC3000_6h_01_eti",
             "DC3000_6h_02_eti",
             "DC3000_6h_03_eti",
             "AvrRpt2_12h_01_eti",
             "AvrRpt2_12h_02_eti",
             "AvrRpt2_12h_03_eti",
             "AvrRpt2_24h_01_eti",
             "AvrRpt2_24h_02_eti",
             "AvrRpt2_24h_03_eti",
             "AvrRpt2_6h_01_eti",
             "AvrRpt2_6h_02_eti",
             "AvrRpt2_6h_03_eti",
             "AvrRpm1_12h_01_eti",
             "AvrRpm1_12h_02_eti",
             "AvrRpm1_12h_03_eti",
             "AvrRpm1_24h_01_eti",
             "AvrRpm1_24h_02_eti",
             "AvrRpm1_24h_03_eti",
             "AvrRpm1_6h_01_eti",
             "AvrRpm1_6h_02_eti",
             "AvrRpm1_6h_03_eti",
             "flg_12h_01_pti",
             "flg_12h_02_pti",
             "flg_12h_03_pti",
             "flg_24h_01_pti",
             "flg_24h_02_pti",
             "flg_24h_03_pti",
             "flg_6h_01_pti",
             "flg_6h_02_pti",
             "flg_6h_03_pti",
             "mock_12h_01_pti",
             "mock_12h_02_pti",
             "mock_12h_03_pti",
             "mock_24h_01_pti",
             "mock_24h_02_pti",
             "mock_24h_03_pti",
             "mock_6h_01_pti",
             "mock_6h_02_pti",
             "mock_6h_03_pti"
              )

#Reading in our files

files <-file.path(list.files("C:/Users/rowbr/OneDrive/Documents/Counts"))

#printing out the file list to makes sure all the files are there

files

#Assigning our sample names to the count files 

names(files)<- samples

#Printing out files again to make sure they all have the approriate names now

files

#Setting up our experimental variables and creating a data frame, with our varaiables being factors with multiple levels:

treatment <- c(rep("mock", 9),
               rep("DC3000", 9),
               rep ("AvrRpt2", 9),
               rep("AvrRpm1", 9),
               rep("flg", 9),
               rep("mock", 9))

treatment <-as.factor(treatment)

length(treatment)

timepoint <- c(rep("6h",3 ),
               rep("12h", 3),
               rep("24h", 3),
               rep("6h", 3),
               rep("12h",3),
               rep("24h", 3),
               rep("6h", 3),
               rep("12h",3),
               rep("24h", 3),
               rep("6h", 3),
               rep("12h",3),
               rep("24h", 3),
               rep("6h", 3),
               rep("12h",3),
               rep("24h", 3),
               rep("6h", 3),
               rep("12h",3),
               rep("24h", 3))
              

timepoint <- as.factor(timepoint)

length(timepoint)

dataset <-c(rep("eti", 36),
            rep("pti", 18))

dataset <-as.factor(dataset)

length(dataset)

# Telling DESeq2 what our experimental design is:

expt.design <- data.frame(sample=samples,
                          file=files,
                          treatment=treatment,
                          timepoint=timepoint,
                          dataset=dataset)

# Double checking our experimental design

expt.design

# First, we're going to assess lab effects - each dataset was generated by a different lab, and so this has the potential to be confounding. 
# Lets create a simple model using only 'dataset', and conduct a Prinicpal component analysis (PCA)

first.model <-formula(~dataset)

# We need to use 'DESeqDataSetFromHTSeqCount' here because our reads were generated with htseq-count

dataset_test <- DESeqDataSetFromHTSeqCount(expt.design, design=first.model)

dataset_test2 <-DESeq(dataset_test)

dataset_test2_results <- results(dataset_test2, alpha=0.05)

# Summarizing the results 

summary(dataset_test2_results)

#Now we can start building the PCA, and in this case, we want to leave it blind because we want to ignore the model information for quality control
#The rlog part adjusts for sample dispersion and library size

pca_test <-rlog((dataset_test2), blind=T)

dim(pca_test)

#Playing around with creating PCA plots

plotPCA(pca_test, intgroup=c("dataset"), ntop=10000)

plotPCA(pca_test, intgroup=c("treatment", "timepoint"), ntop=10000)

plotPCA(pca_test, intgroup=c("treatment", "timepoint", "dataset"), ntop=10000)

# Now for the actual differential expression analysis. I spent a long time thinking about how to set up the model and I'm still not entirely sure if this is the most correct way to do it.
# I ignored the timepoints in favor of a simpler initial analysis
# I think it would be interesting to come back to the timepoints though and build them into a more complex model later

load.model <-formula(~dataset+treatment)

countData <-DESeqDataSetFromHTSeqCount(expt.design, design=load.model)

# Telling DESeq2 what the different treatment groups are

countData$group <-factor(paste(countData$treatment, countData$dataset, sep="_"))
countData$group = factor(countData$group, levels=c("mock_eti", "mock_pti","DC3000_eti","AvrRpm1_eti", "AvrRpt2_eti", "flg_pti"))

# Filtering out low count genes. This was reccomended in the DESeq2 vignette as it can help reduce the memory size of the object and help things go a bit faster
# Garrett also used this filtering out step

keep<-rowMeans(counts(countData))>=10
countData<-countData[keep,]

countData@design=~group

# Actualy doing the differential expression analysis with 'DESeq'

countData2<-DESeq(countData)

# Checking out our differential expression analysis results

res <-results(countData2)

res 

summary(res)

------------------------------------------------------------------------------------------------------------------------------------------

#Performing the contrast analyses on different pairs of treatment/dataset groups

pti_vs_eti_mock=results(countData2, contrast=c("group", "mock_eti", "mock_pti"), alpha=0.05)
summary (pti_vs_eti_mock)

pti_vs_eti_mock=pti_vs_eti_mock[is.na(pti_vs_eti_mock$padj)==F,]
pti_vs_eti_mock=pti_vs_eti_mock[pti_vs_eti_mock$log2FoldChange >1 & pti_vs_eti_mock$padj <0.05,]

flg_vs_AvrRpt=results(countData2, contrast = c("group", "flg_pti", "AvrRpt2_eti"), alpha = 0.05)

flg_vs_AvrRpt=flg_vs_AvrRpt[is.na(flg_vs_AvrRpt$padj)==F,]
flg_vs_AvrRpt=flg_vs_AvrRpt[flg_vs_AvrRpt$log2FoldChange >1 & flg_vs_AvrRpt$padj <0.05,]

flg_vs_mock =results(countData2, contrast = c("group", "flg_pti", "mock_pti"), alpha = 0.05)

flg_vs_mock=flg_vs_mock[is.na(flg_vs_mock$padj)==F,]
flg_vs_mock=flg_vs_mock[flg_vs_mock$log2FoldChange >1 & flg_vs_mock$padj <0.05,]

flg_vs_dc3000=results(countData2, contrast= c("group", "flg_pti", "DC3000_eti"), alpha = 0.05)

flg_vs_dc3000=flg_vs_dc3000[is.na(flg_vs_dc3000$padj)==F,]
flg_vs_dc3000=flg_vs_dc3000[flg_vs_dc3000$log2FoldChange >1 & flg_vs_dc3000$padj <0.05,]

flg_vs_AvrRpm1 = results(countData2, contrast = c("group", "flg_pti", "AvrRpm1_eti"), alpha=0.05)

flg_vs_AvrRpm1=flg_vs_AvrRpm1[is.na(flg_vs_AvrRpm1$padj)==F,]
flg_vs_AvrRpm1=flg_vs_AvrRpm1[flg_vs_AvrRpm1$log2FoldChange >1 & flg_vs_AvrRpm1$padj <0.05,]

mock_vs_dc3000=results(countData2, contrast = c("group", "mock_eti", "DC3000_eti"), alpha=0.05)

mock_vs_dc3000=mock_vs_dc3000[is.na(mock_vs_dc3000$padj)==F,]
mock_vs_dc3000=mock_vs_dc3000[mock_vs_dc3000$log2FoldChange > 1 & mock_vs_dc3000$padj <0.05,]

mock_vs_AvrRpt2=results(countData2, contrast=c("group", "mock_eti", "AvrRpt2_eti"), alpha=0.05)

mock_vs_AvrRpt2=mock_vs_AvrRpt2[is.na(mock_vs_AvrRpt2$padj)==F,]
mock_vs_AvrRpt2=mock_vs_AvrRpt2[mock_vs_AvrRpt2$log2FoldChange >1 & mock_vs_AvrRpt2$padj <0.05,]

mock_vs_AvrRpm1=results(countData2, contrast=c("group", "mock_eti", "AvrRpm1_eti"), alpha=0.05)

mock_vs_AvrRpm1=mock_vs_AvrRpm1[is.na(mock_vs_AvrRpm1$padj)==F,]
mock_vs_AvrRpm1=mock_vs_AvrRpm1[mock_vs_AvrRpm1$log2FoldChange >1 & mock_vs_AvrRpm1$padj <0.05,]

----------------------------------------------------------------------------------------------------------------------------------
#Ordering the contrast outputs by padj and then writing them to csv so I can save them to my laptop/drive

head (pti_vs_eti_mock)

write.csv(pti_vs_eti_mock, "pti_vs_eti_mock")

pti_vs_eti_mock<-pti_vs_eti_mock[order(pti_vs_eti_mock$padj),]

             
head (pti_vs_eti_mock)

flg_vs_AvrRpt <-flg_vs_AvrRpt[order(flg_vs_AvrRpt$padj),]
head (flg_vs_AvrRpt)

write.csv(flg_vs_AvrRpt, "flg_vs_AvrRpt2")

----------------------------------------------------------------------------------------------------------------------------------

# My attempt at making a hierarchical clustering heatmap from the code we used in class - I wasn't happy with how mine looked so I cut it from the report. It felt very overcrowded with all the different sample names.

rlogMat <-assay(pca_test)

distsRL <-dist(t(rlogMat))

mat <-as.matrix(distsRL)

rownames(mat)<- colnames(mat) <-with(colData(countData2), paste(treatment,paste0(dataset)))

hc <-hclust(distsRL)

hmcol <-colorRampPalette(brewer.pal(9, "Blues"))(100) # some of these colour palettes were nice to play around with though!

heatmap.2(mat, Rowv=as.dendrogram(hc), symm = TRUE, trace = "none", col=rev(hmcol), margin=c(10,10))
